ext/node/path/mod.rs

```diff
@@ -38,3 +46,33 @@ impl PathClean<PathBuf> for PathBuf {
     }
   }
 }
+
+pub fn is_windows_device_root(code: u32) -> bool {
+  (constants::CHAR_LOWERCASE_A..=constants::CHAR_LOWERCASE_Z).contains(&code)
+    || (constants::CHAR_UPPERCASE_A..=constants::CHAR_UPPERCASE_Z)
+      .contains(&code)
+}
+
+pub fn is_posix_path_separator(code: u32) -> bool {
+  code == constants::CHAR_FORWARD_SLASH
+}
+
+pub fn is_path_separator(code: u32) -> bool {
+  is_posix_path_separator(code) || code == constants::CHAR_BACKWARD_SLASH
+}
+
+pub fn char_at_index(path: &str, index: usize) -> Option<char> {
+  path.chars().nth(index)
+}
+
+pub fn assert_path(path: &Value) -> Result<&str, AnyError> {
+  match path {
+    Value::String(p) => Ok(p),
+    _ => {
+      let msg = format!(
+        "[ERR_INVALID_ARG_TYPE] The \"path\" argument must be of type string. Received type {path:?}",
+      );
+      Err(type_error(msg))
+    }
+  }
+}
```

Comment:
@bartlomieju do errors returned in Rust not have codes in the exceptions thrown in Javascript userland ? 

I have been trying to get this assert_path function to return an Err so in userland the exception has a code. This is for the [test-path.js](https://github.com/denoland/deno/blob/main/cli/tests/node_compat/test/parallel/test-path.js#L34-L43) tests.

I doubt am I suppose to catch the error in JS userland and than throw `throw new ERR_INVALID_ARG_TYPE("path", ["string"], path);`.

Reply:
With this assert_path function the exception thrown is + console.log(err.code) after:

```bash
# ./target/debug/deno run /tmp/deno_tests/file.ts
TypeError: [ERR_INVALID_ARG_TYPE] The "path" argument must be of type string. Received type Number(7)
    at Object.dirname (internal:deno_node/polyfills/path/posix.ts:173:16)
    at file:///tmp/deno_tests/file.ts:7:14
exception code:  undefined
```

This output is similar to how NodeJS logs:

```
node /tmp/deno_tests/file.ts
TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received type number (7)
    at new NodeError (node:internal/errors:371:5)
    at validateString (node:internal/validators:119:11)
    at Object.dirname (node:path:1276:5)
    at Object.<anonymous> (/tmp/deno_tests/file.ts:7:14)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:17:47 {
  code: 'ERR_INVALID_ARG_TYPE'
}
exception code:  ERR_INVALID_ARG_TYPE
```

Deno is not like Node and has a random " also.. The code is handled correctly though since this exception is thrown in Javascript userland.

```bash
deno run /tmp/deno_tests/file.ts
TypeError [ERR_INVALID_ARG_TYPE]" argument must be of type string. Received type number (7)
    at assertPath (https://deno.land/std@0.176.0/node/path/_util.ts:19:11)
    at Object.dirname (https://deno.land/std@0.176.0/node/path/posix.ts:223:3)
    at file:///tmp/deno_tests/file.ts:7:14
exception code:  ERR_INVALID_ARG_TYPE
```

Reply:
@20k-ultra your solution is fine for now - we already do the same thing in `ext/node/errors.rs`

Reply:
I actually had my error in that file but than thought... Deno must have an error in deno_core to say an invalid arg type was passed and this isn't specific to node.

So do you think I should move my error to errors.rs, catch it in JS userland, check error.message.starts_with("[ERR_INVALID_ARG_TYPE]" and `throw ERR_INVALID_ARG_TYPE("path", ["string"], path);` ?